pkg/queue/sharedmain/main.go

```diff
@@ -306,15 +309,36 @@ func Main(opts ...Option) error {
 		logger.Infof("Sleeping %v to allow K8s propagation of non-ready state", drainSleepDuration)
 		drainer.Drain()
 
+		ctx, cancel := context.WithTimeout(context.Background(), time.Duration(env.RevisionTimeoutSeconds)*time.Second)
```

Comment:
If there was a long running request close to the revision timeout we already have waited that long (in Serving we set terminationGracePeriod = revision_timeout). We should only wait for the quiet time (which we have done in the drainer already again).  You just want to exit not wait in the server shutdown given that K8s [will not wait anyway](https://www.cncf.io/blog/2024/12/19/decoding-the-pod-termination-lifecycle-in-kubernetes-a-comprehensive-guide).

How do you make sure that the pending websocket connection is closed when server.shutdown is not counting that. Does `defer cancel()` actually do the trick? In the [patch](https://gist.github.com/skonto/2fef5ce5bf10b48358eb61e1d9879897) I pasted I was hoping that the cancelled request context would kill the webhook connection.

Reply:
Would you not want some sort of exit condition? I can remove it, I just added as an explicit safeguard (as opposed to implicitly relying on the revision). I do not rely on the context to cancel properly. It is only there for the timeout to make sure the wait exits. I did try the patch, it did not work. `defer cancel()` does not work by itself to ensure graceful termination. 

With the pending request counter, you do not need to check the websocket connection explicitly. When the handler terminates, caused by the websocket closing, the counter will go down. This is essentially just implementing a pre-shutdown where we don't rely on `.Shutdown()` to wait on requests to finish which will drop the highjacked ws connections.

Reply:
> Yes but it is misleading, you don't have that time left.
I get what you're saying that functionally it doesn't matter cause of eventual sigkill. Surely you do have approximately that much time left to shut down (I guess minus the drain time)? Seems more explicit to include. Regardless, I will remove the context

Yeah I did try the explicit cancel you described. It did not work. The issue is with `.Shutdown`'s behaviour..